pkg/controllers/deprovisioning/controller.go

```diff
@@ -209,26 +209,27 @@ func (c *Controller) executeCommand(ctx context.Context, d Deprovisioner, comman
 // nolint:gocyclo
 func (c *Controller) launchReplacementMachines(ctx context.Context, action Command, reason string) error {
 	defer metrics.Measure(deprovisioningReplacementNodeInitializedHistogram)()
-	candidateNodeNames := lo.Map(action.candidates, func(c *Candidate, _ int) string { return c.Node.Name })
 
 	// cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
-	if err := c.setNodesUnschedulable(ctx, true, candidateNodeNames...); err != nil {
+	if err := c.setNodesUnschedulable(ctx, true, action.candidates...); err != nil {
 		return fmt.Errorf("cordoning nodes, %w", err)
 	}
 
 	nodeClaimKeys, err := c.provisioner.CreateNodeClaims(ctx, action.replacements, provisioning.WithReason(reason))
 	if err != nil {
 		// uncordon the nodes as the launch may fail (e.g. ICE)
-		err = multierr.Append(err, c.setNodesUnschedulable(ctx, false, candidateNodeNames...))
+		err = multierr.Append(err, c.setNodesUnschedulable(ctx, false, action.candidates...))
 		return err
 	}
 	if len(nodeClaimKeys) != len(action.replacements) {
 		// shouldn't ever occur since a partially failed CreateNodeClaims should return an error
 		return fmt.Errorf("expected %d nodes, got %d", len(action.replacements), len(nodeClaimKeys))
 	}
 
+	candidateProviderIDs := lo.Map(action.candidates, func(c *Candidate, _ int) string { return c.ProviderID() })
+
 	// We have the new machines created at the API server so mark the old machines for deletion
-	c.cluster.MarkForDeletion(candidateNodeNames...)
+	c.cluster.MarkForDeletion(candidateProviderIDs...)
```

Comment:
does changing to provider ID affect a node that never gets a provider ID for some reason?